File: src/web/static/js/validation/unified-validation-system.js
/**
 * Unified Validation System - V2 Compliant
 * Consolidates all duplicate validation functions across web development modules
 * 
 * @author Agent-7 - Web Development Specialist
 * @version 1.0.0 - V2 COMPLIANCE CONSOLIDATION
 * @license MIT
 */

// ================================
// UNIFIED VALIDATION SYSTEM
// ================================

/**
 * Unified validation system that consolidates all duplicate validation functions
 * from dashboard validation utilities, utility validation service, and dashboard utils
 */
class UnifiedValidationSystem {
    constructor() {
        this.logger = console;
        this.cache = new Map();
    }

    /**
     * Validate email address - consolidated from multiple sources
     */
    validateEmail(email) {
        try {
            if (!email || typeof email !== 'string') {
                return { isValid: false, error: 'Email must be a valid string' };
            }

            const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
            const isValid = emailRegex.test(email);

            return {
                isValid: isValid,
                error: isValid ? null : 'Invalid email format'
            };
        } catch (error) {
            this.logError('Email validation failed', error);
            return { isValid: false, error: 'Email validation error' };
        }
    }

    /**
     * Validate URL - consolidated from utility validation service
     */
    validateUrl(url) {
        try {
            if (!url || typeof url !== 'string') {
                return { isValid: false, error: 'URL must be a valid string' };
            }

            try {
                new URL(url);
                return { isValid: true, error: null };
            } catch (urlError) {
                return { isValid: false, error: 'Invalid URL format' };
            }
        } catch (error) {
            this.logError('URL validation failed', error);
            return { isValid: false, error: 'URL validation error' };
        }
    }

    /**
     * Validate required fields in object - consolidated from utility validation service
     */
    validateRequiredFields(data, requiredFields) {
        try {
            if (!data || typeof data !== 'object') {
                return {
                    isValid: false,
                    missingFields: requiredFields,
                    error: 'Data object is required'
                };
            }

            if (!Array.isArray(requiredFields)) {
                return {
                    isValid: false,
                    missingFields: [],
                    error: 'Required fields must be an array'
                };
            }

            const missingFields = [];
            const invalidFields = [];

            for (const field of requiredFields) {
                if (!(field in data)) {
                    missingFields.push(field);
                } else if (data[field] === null || data[field] === undefined || data[field] === '') {
                    invalidFields.push(field);
                }
            }

            const isValid = missingFields.length === 0 && invalidFields.length === 0;

            return {
                isValid: isValid,
                missingFields: missingFields,
                invalidFields: invalidFields,
                error: isValid ? null : `Missing: ${missingFields.join(', ')}, Invalid: ${invalidFields.join(', ')}`
            };
        } catch (error) {
            this.logError('Required fields validation failed', error);
            return {
                isValid: false,
                missingFields: requiredFields,
                error: 'Required fields validation error'
            };
        }
    }

    /**
     * Validate numeric range - consolidated from utility validation service
     */
    validateNumericRange(value, min = null, max = null) {
        try {
            const numValue = Number(value);
            
            if (isNaN(numValue) || !isFinite(numValue)) {
                return { isValid: false, error: 'Value must be a valid number' };
            }

            if (min !== null && numValue < min) {
                return { isValid: false, error: `Value must be at least ${min}` };
            }

            if (max !== null && numValue > max) {
                return { isValid: false, error: `Value must be at most ${max}` };
            }

            return { isValid: true, error: null };
        } catch (error) {
            this.logError('Numeric range validation failed', error);
            return { isValid: false, error: 'Numeric range validation error' };
        }
    }

    /**
     * Validate dashboard configuration - consolidated from dashboard validation utilities
     */
    validateDashboardConfig(config) {
        try {
            if (!config || typeof config !== 'object') {
                return { isValid: false, error: 'Dashboard configuration must be a valid object' };
            }

            const requiredFields = ['id', 'title'];
            const missingFields = requiredFields.filter(field => !config[field]);

            if (missingFields.length > 0) {
                return {
                    isValid: false,
                    error: `Dashboard configuration missing required fields: ${missingFields.join(', ')}`
                };
            }

            return { isValid: true, error: null };
        } catch (error) {
            this.logError('Dashboard config validation failed', error);
            return { isValid: false, error: 'Dashboard config validation error' };
        }
    }

    /**
     * Validate chart data - consolidated from dashboard validation utilities
     */
    validateChartData(data) {
        try {
            if (!data || !Array.isArray(data)) {
                return { isValid: false, error: 'Chart data must be a valid array' };
            }

            if (data.length === 0) {
                return { isValid: false, error: 'Chart data cannot be empty' };
            }

            // Validate data structure
            const requiredFields = ['label', 'value'];
            for (let i = 0; i < data.length; i++) {
                const item = data[i];
                if (!item || typeof item !== 'object') {
                    return { isValid: false, error: `Chart data item ${i} must be an object` };
                }

                const missingFields = requiredFields.filter(field => !(field in item));
                if (missingFields.length > 0) {
                    return {
                        isValid: false,
                        error: `Chart data item ${i} missing required fields: ${missingFields.join(', ')}`
                    };
                }
            }

            return { isValid: true, error: null };
        } catch (error) {
            this.logError('Chart data validation failed', error);
            return { isValid: false, error: 'Chart data validation error' };
        }
    }

    /**
     * Validate API response - consolidated from dashboard validation utilities
     */
    validateApiResponse(response, expectedStructure = {}) {
        try {
            if (!response || typeof response !== 'object') {
                return { isValid: false, error: 'API response must be a valid object' };
            }

            // Check expected structure if provided
            if (Object.keys(expectedStructure).length > 0) {
                for (const [key, expectedType] of Object.entries(expectedStructure)) {
                    if (!(key in response)) {
                        return { isValid: false, error: `API response missing required field: ${key}` };
                    }

                    if (typeof response[key] !== expectedType) {
                        return { isValid: false, error: `API response field ${key} must be of type ${expectedType}` };
                    }
                }
            }

            return { isValid: true, error: null };
        } catch (error) {
            this.logError('API response validation failed', error);
            return { isValid: false, error: 'API response validation error' };
        }
    }

    /**
     * Validate form input - consolidated from dashboard validation utilities
     */
    validateFormInput(value, rules = {}) {
        try {
            const errors = [];

            // Required validation
            if (rules.required && (!value || (typeof value === 'string' && value.trim() === ''))) {
                errors.push('This field is required');
            }

            // Type validation
            if (rules.type && typeof value !== rules.type) {
                errors.push(`Must be of type ${rules.type}`);
            }

            // Length validation
            if (rules.minLength && typeof value === 'string' && value.length < rules.minLength) {
                errors.push(`Must be at least ${rules.minLength} characters`);
            }

            if (rules.maxLength && typeof value === 'string' && value.length > rules.maxLength) {
                errors.push(`Must be at most ${rules.maxLength} characters`);
            }

            // Pattern validation
            if (rules.pattern && typeof value === 'string' && !rules.pattern.test(value)) {
                errors.push('Invalid format');
            }

            return {
                isValid: errors.length === 0,
                error: errors.length > 0 ? errors[0] : null,
                allErrors: errors
            };
        } catch (error) {
            this.logError('Form input validation failed', error);
            return {
                isValid: false,
                error: 'Form input validation error',
                allErrors: ['Form input validation error']
            };
        }
    }

    /**
     * Type validation utilities - consolidated from dashboard utils
     */
    createTypeValidationUtils() {
        return {
            isEmail: (email) => this.validateEmail(email).isValid,
            isNumber: (value) => !isNaN(value) && isFinite(value),
            isString: (value) => typeof value === 'string',
            isObject: (value) => typeof value === 'object' && value !== null,
            isArray: (value) => Array.isArray(value),
            isBoolean: (value) => typeof value === 'boolean'
        };
    }

    /**
     * Log error with context
     */
    logError(message, error) {
        this.logger.error(`[UnifiedValidationSystem] ${message}:`, error);
    }

    /**
     * Clear validation cache
     */
    clearCache() {
        this.cache.clear();
    }
}

// ================================
// FACTORY FUNCTIONS
// ================================

/**
 * Create unified validation system instance
 */
export function createUnifiedValidationSystem() {
    return new UnifiedValidationSystem();
}

/**
 * Create type validation utilities
 */
export function createTypeValidationUtils() {
    const system = new UnifiedValidationSystem();
    return system.createTypeValidationUtils();
}

// ================================
// EXPORTS
// ================================

export { UnifiedValidationSystem };
export default UnifiedValidationSystem;

File: src/core/unified_validation_system.py
#!/usr/bin/env python3
"""
Unified Validation System - DRY Violation Elimination
====================================================

Eliminates duplicate validation patterns across 8+ modules:
- src/core/validation/coordination_validator.py
- src/utils/config_pattern_scanner.py
- src/services/messaging_core.py
- src/core/validation/javascript_v2_compliance_validator.py
- src/core/validation/agent7_v2_compliance_coordinator.py
- src/core/validation/repository_pattern_validator.py
- src/core/validation/enhanced_cli_validation_framework.py
- src/core/validation/gaming_performance_threshold_checker.py

CONSOLIDATED: Single source of truth for all validation operations.

Author: Agent-5 (Business Intelligence Specialist)
Mission: DRY Violation Elimination
Status: CONSOLIDATED - Validation Duplication Eliminated
"""

import re
from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import Enum
from typing import Any, Dict, List, Optional, Union, Callable

# Removed circular import - will use local fallbacks


# ================================
# VALIDATION TYPES
# ================================

class ValidationType(Enum):
    """Types of validation operations."""
    
    REQUIRED_FIELDS = "required_fields"
    DATA_TYPES = "data_types"
    EMAIL = "email"
    URL = "url"
    STRING_LENGTH = "string_length"
    NUMERIC_RANGE = "numeric_range"
    REGEX_PATTERN = "regex_pattern"
    CUSTOM = "custom"

class ValidationSeverity(Enum):
    """Validation severity levels."""
    
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"
    INFO = "info"

# ================================
# VALIDATION RESULT
# ================================

@dataclass
class ValidationResult:
    """Result of a validation operation."""
    
    is_valid: bool
    errors: List[str] = None
    warnings: List[str] = None
    severity: ValidationSeverity = ValidationSeverity.MEDIUM
    field_name: Optional[str] = None
    validation_type: Optional[ValidationType] = None
    details: Optional[Dict[str, Any]] = None
    
    def __post_init__(self):
        if self.errors is None:
            self.errors = []
        if self.warnings is None:
            self.warnings = []
        if self.details is None:
            self.details = {}

# ================================
# VALIDATION RULES
# ================================

@dataclass
class ValidationRule:
    """Validation rule definition."""
    
    rule_type: ValidationType
    field_name: str
    parameters: Dict[str, Any]
    severity: ValidationSeverity = ValidationSeverity.MEDIUM
    custom_validator: Optional[Callable] = None
    error_message: Optional[str] = None

# ================================
# UNIFIED VALIDATION SYSTEM
# ================================

class UnifiedValidationSystem:
    """
    Unified validation system that eliminates duplicate validation patterns.
    
    CONSOLIDATES:
    - validate_required_fields(data, required_fields) (duplicated in 8+ files)
    - validate_data_types(data, type_requirements) (duplicated in 8+ files)
    - validate_email (duplicated in 5+ files)
    - validate_url (duplicated in 4+ files)
    - validate_string_length (duplicated in 6+ files)
    - validate_numeric_range (duplicated in 3+ files)
    - validate_regex_pattern (duplicated in 4+ files)
    """
    
    def __init__(self):
        """Initialize unified validation system."""
        self.config = {}
        import logging
        self.logger = logging.getLogger(__name__)
        
        # Validation patterns
        self._patterns = {
            "email": re.compile(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'),
            "url": re.compile(r'^https?://(?:[-\w.])+(?:\:[0-9]+)?(?:/(?:[\w/_.])*(?:\?(?:[\w&=%.])*)?(?:\#(?:[\w.])*)?)?$'),
            "phone": re.compile(r'^\+?1?[-.\s]?\(?([0-9]{3})\)?[-.\s]?([0-9]{3})[-.\s]?([0-9]{4})$'),
            "uuid": re.compile(r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$', re.IGNORECASE),
            "alphanumeric": re.compile(r'^[a-zA-Z0-9]+$'),
            "numeric": re.compile(r'^[0-9]+$'),
            "alpha": re.compile(r'^[a-zA-Z]+$')
        }
        
        # Default validation rules
        self._default_rules = {
            "string_length": {"min": 0, "max": 1000},
            "numeric_range": {"min": 0, "max": 999999},
            "required_fields": [],
            "data_types": {}
        }
    
    def validate_required(self, data: Any) -> bool:
        """Validate that data is not None or empty."""
        return data is not None and data != "" and data != []
    
    def validate_required_fields(self, data: Dict[str, Any], required_fields: List[str]) -> ValidationResult:
        """
        Validate required fields - consolidates 8+ duplicate implementations.
        
        CONSOLIDATES FROM:
        - src/core/validation/coordination_validator.py
        - src/utils/config_pattern_scanner.py
        - src/services/messaging_core.py
        - src/core/validation/javascript_v2_compliance_validator.py
        """
        missing_fields = []
        
        for field in required_fields:
            if field not in data or data[field] is None or data[field] == "":
                missing_fields.append(field)
        
        is_valid = len(missing_fields) == 0
        errors = [f"Required field missing: {field}" for field in missing_fields] if missing_fields else []
        
        return ValidationResult(
            is_valid=is_valid,
            errors=errors,
            severity=ValidationSeverity.HIGH,
            validation_type=ValidationType.REQUIRED_FIELDS,
            details={"missing_fields": missing_fields}
        )
    
    def validate_data_types(self, data: Dict[str, Any], type_requirements: Dict[str, type]) -> ValidationResult:
        """
        Validate data types - consolidates 8+ duplicate implementations.
        
        CONSOLIDATES FROM:
        - src/core/validation/coordination_validator.py
        - src/utils/config_pattern_scanner.py
        - src/services/messaging_core.py
        - src/core/validation/javascript_v2_compliance_validator.py
        """
        invalid_fields = []
        
        for field, expected_type in type_requirements.items():
            if field in data and not isinstance(data[field], expected_type):
                invalid_fields.append(field)
        
        is_valid = len(invalid_fields) == 0
        errors = [f"Invalid type for field '{field}': expected {type_requirements[field].__name__}" 
                 for field in invalid_fields] if invalid_fields else []
        
        return ValidationResult(
            is_valid=is_valid,
            errors=errors,
            severity=ValidationSeverity.MEDIUM,
            validation_type=ValidationType.DATA_TYPES,
            details={"invalid_fields": invalid_fields}
        )
    
    def validate_email(self, email: str) -> ValidationResult:
        """Validate email format - consolidates 5+ duplicate implementations."""
        if not email:
            return ValidationResult(
                is_valid=False,
                errors=["Email is required"],
                severity=ValidationSeverity.HIGH,
                validation_type=ValidationType.EMAIL
            )
        
        is_valid = bool(self._patterns["email"].match(email))
        errors = ["Invalid email format"] if not is_valid else []
        
        return ValidationResult(
            is_valid=is_valid,
            errors=errors,
            severity=ValidationSeverity.MEDIUM,
            validation_type=ValidationType.EMAIL
        )
    
    def validate_url(self, url: str) -> ValidationResult:
        """Validate URL format - consolidates 4+ duplicate implementations."""
        if not url:
            return ValidationResult(
                is_valid=False,
                errors=["URL is required"],
                severity=ValidationSeverity.HIGH,
                validation_type=ValidationType.URL
            )
        
        is_valid = bool(self._patterns["url"].match(url))
        errors = ["Invalid URL format"] if not is_valid else []
        
        return ValidationResult(
            is_valid=is_valid,
            errors=errors,
            severity=ValidationSeverity.MEDIUM,
            validation_type=ValidationType.URL
        )
    
    def validate_string_length(self, value: str, min_length: int = 0, max_length: int = 1000) -> ValidationResult:
        """Validate string length - consolidates 6+ duplicate implementations."""
        if not isinstance(value, str):
            return ValidationResult(
                is_valid=False,
                errors=["Value must be a string"],
                severity=ValidationSeverity.HIGH,
                validation_type=ValidationType.STRING_LENGTH
            )
        
        length = len(value)
        errors = []
        
        if length < min_length:
            errors.append(f"String too short: {length} < {min_length}")
        if length > max_length:
            errors.append(f"String too long: {length} > {max_length}")
        
        is_valid = len(errors) == 0
        
        return ValidationResult(
            is_valid=is_valid,
            errors=errors,
            severity=ValidationSeverity.MEDIUM,
            validation_type=ValidationType.STRING_LENGTH,
            details={"length": length, "min_length": min_length, "max_length": max_length}
        )
    
    def validate_numeric_range(self, value: Union[int, float], min_value: float = 0, max_value: float = 999999) -> ValidationResult:
        """Validate numeric range - consolidates 3+ duplicate implementations."""
        if not isinstance(value, (int, float)):
            return ValidationResult(
                is_valid=False,
                errors=["Value must be numeric"],
                severity=ValidationSeverity.HIGH,
                validation_type=ValidationType.NUMERIC_RANGE
            )
        
        errors = []
        
        if value < min_value:
            errors.append(f"Value too small: {value} < {min_value}")
        if value > max_value:
            errors.append(f"Value too large: {value} > {max_value}")
        
        is_valid = len(errors) == 0
        
        return ValidationResult(
            is_valid=is_valid,
            errors=errors,
            severity=ValidationSeverity.MEDIUM,
            validation_type=ValidationType.NUMERIC_RANGE,
            details={"value": value, "min_value": min_value, "max_value": max_value}
        )
    
    def validate_regex_pattern(self, value: str, pattern_name: str) -> ValidationResult:
        """Validate regex pattern - consolidates 4+ duplicate implementations."""
        if not isinstance(value, str):
            return ValidationResult(
                is_valid=False,
                errors=["Value must be a string"],
                severity=ValidationSeverity.HIGH,
                validation_type=ValidationType.REGEX_PATTERN
            )
        
        if pattern_name not in self._patterns:
            return ValidationResult(
                is_valid=False,
                errors=[f"Unknown pattern: {pattern_name}"],
                severity=ValidationSeverity.HIGH,
                validation_type=ValidationType.REGEX_PATTERN
            )
        
        is_valid = bool(self._patterns[pattern_name].match(value))
        errors = [f"Value does not match pattern: {pattern_name}"] if not is_valid else []
        
        return ValidationResult(
            is_valid=is_valid,
            errors=errors,
            severity=ValidationSeverity.MEDIUM,
            validation_type=ValidationType.REGEX_PATTERN,
            details={"pattern_name": pattern_name}
        )
    
    def validate_custom(self, value: Any, validator: Callable, field_name: str = "field") -> ValidationResult:
        """Validate using custom validator function."""
        try:
            result = validator(value)
            if isinstance(result, bool):
                is_valid = result
                errors = [] if is_valid else [f"Custom validation failed for {field_name}"]
            elif isinstance(result, ValidationResult):
                return result
            else:
                is_valid = bool(result)
                errors = [] if is_valid else [f"Custom validation failed for {field_name}"]
        except Exception as e:
            is_valid = False
            errors = [f"Custom validation error for {field_name}: {str(e)}"]
        
        return ValidationResult(
            is_valid=is_valid,
            errors=errors,
            severity=ValidationSeverity.MEDIUM,
            validation_type=ValidationType.CUSTOM,
            field_name=field_name
        )
    
    def validate_multiple(self, data: Dict[str, Any], rules: List[ValidationRule]) -> List[ValidationResult]:
        """Validate multiple fields with different rules."""
        results = []
        
        for rule in rules:
            if rule.field_name not in data:
                if rule.severity == ValidationSeverity.HIGH:
                    results.append(ValidationResult(
                        is_valid=False,
                        errors=[f"Required field missing: {rule.field_name}"],
                        severity=rule.severity,
                        field_name=rule.field_name,
                        validation_type=rule.rule_type
                    ))
                continue
            
            value = data[rule.field_name]
            
            if rule.rule_type == ValidationType.REQUIRED_FIELDS:
                result = self.validate_required_fields({rule.field_name: value}, [rule.field_name])
            elif rule.rule_type == ValidationType.DATA_TYPES:
                result = self.validate_data_types({rule.field_name: value}, {rule.field_name: rule.parameters.get("type", str)})
            elif rule.rule_type == ValidationType.EMAIL:
                result = self.validate_email(value)
            elif rule.rule_type == ValidationType.URL:
                result = self.validate_url(value)
            elif rule.rule_type == ValidationType.STRING_LENGTH:
                result = self.validate_string_length(value, rule.parameters.get("min", 0), rule.parameters.get("max", 1000))
            elif rule.rule_type == ValidationType.NUMERIC_RANGE:
                result = self.validate_numeric_range(value, rule.parameters.get("min", 0), rule.parameters.get("max", 999999))
            elif rule.rule_type == ValidationType.REGEX_PATTERN:
                result = self.validate_regex_pattern(value, rule.parameters.get("pattern", "alphanumeric"))
            elif rule.rule_type == ValidationType.CUSTOM:
                result = self.validate_custom(value, rule.custom_validator, rule.field_name)
            else:
                result = ValidationResult(
                    is_valid=False,
                    errors=[f"Unknown validation type: {rule.rule_type}"],
                    severity=ValidationSeverity.HIGH,
                    field_name=rule.field_name,
                    validation_type=rule.rule_type
                )
            
            result.field_name = rule.field_name
            result.validation_type = rule.rule_type
            result.severity = rule.severity
            results.append(result)
        
        return results

# ================================
# GLOBAL VALIDATION INSTANCE
# ================================

# Single global instance to eliminate duplicate validation objects
_unified_validator = None

def get_unified_validator() -> UnifiedValidationSystem:
    """Get global unified validation instance."""
    global _unified_validator
    if _unified_validator is None:
        _unified_validator = UnifiedValidationSystem()
    return _unified_validator

# ================================
# CONVENIENCE FUNCTIONS
# ================================

def validate_required_fields(data: Dict[str, Any], required_fields: List[str]) -> ValidationResult:
    """Validate required fields."""
    return get_unified_validator().validate_required_fields(data, required_fields)

def validate_data_types(data: Dict[str, Any], type_requirements: Dict[str, type]) -> ValidationResult:
    """Validate data types."""
    return get_unified_validator().validate_data_types(data, type_requirements)

def validate_email(email: str) -> ValidationResult:
    """Validate email format."""
    return get_unified_validator().validate_email(email)

def validate_url(url: str) -> ValidationResult:
    """Validate URL format."""
    return get_unified_validator().validate_url(url)

def validate_string_length(value: str, min_length: int = 0, max_length: int = 1000) -> ValidationResult:
    """Validate string length."""
    return get_unified_validator().validate_string_length(value, min_length, max_length)

def validate_numeric_range(value: Union[int, float], min_value: float = 0, max_value: float = 999999) -> ValidationResult:
    """Validate numeric range."""
    return get_unified_validator().validate_numeric_range(value, min_value, max_value)

def validate_regex_pattern(value: str, pattern_name: str) -> ValidationResult:
    """Validate regex pattern."""
    return get_unified_validator().validate_regex_pattern(value, pattern_name)

def validate_custom(value: Any, validator: Callable, field_name: str = "field") -> ValidationResult:
    """Validate using custom validator function."""
    return get_unified_validator().validate_custom(value, validator, field_name)
